pkg/cloudprovider/openstack.go

```diff
@@ -237,45 +219,126 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
 				continue
 			}
 			if matchingSubnet != nil {
-				return fmt.Errorf("requested IP address %s for node %s and port %s matches 2 different subnets, %s and %s",
+				return nil, nil, fmt.Errorf("requested IP address %s for node %s and port %s matches 2 different subnets, %s and %s",
 					ip, node.Name, serverPort.ID, matchingSubnet.ID, s.ID)
 			}
 			matchingSubnet = &s
 		}
 		if matchingSubnet != nil {
-			// 2) Reserve the IP address on the subnet by creating a new unattached neutron port.
-			unboundPort, err := o.reserveNeutronIPAddress(*matchingSubnet, ip, serverID)
-			if err != nil {
-				return err
+			// 2) Return the matching subnet and port.
+			return matchingSubnet, &serverPort, nil
+		}
+	}
+
+	// 3) The IP address does not fit in any of the attached networks' subnets.
+	return nil, nil, fmt.Errorf("could not assign IP address %s to node %s", ip, node.Name)
+}
+
+// AssignPrivateIP attempts to assigning the IP address provided to the VM
+// instance corresponding to the corev1.Node provided on the cloud the
+// cluster is deployed on.
+// NOTE: This operation is performed against all interfaces that are attached
+// to the server. In case that an instance has 2 interfaces with the same CIDR
+// that this IP address could fit in, the first interface that is found will be used.
+// No guarantees about the correct interface ordering are given in such a case.
+// Throw an AlreadyExistingIPError if the IP provided is already associated with the
+// node, it's up to the caller to decide what to do with that.
+// NOTE: For OpenStack, this is a 2 step operation which is not atomic:
+//
+//	a) Reserve a neutron port.
+//	b) Add the IP address to the allowed_address_pairs field.
+//
+// If step b) fails, then we will try to undo step a). However, if this undo fails,
+// then we will be in a situation where the user or an upper layer will have to call
+// ReleasePrivateIP to get out of this situation.
+func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	if node == nil {
+		return fmt.Errorf("invalid nil pointer provided for node when trying to assign private IP %s", ip.String())
+	}
+	serverID, err := getNovaServerIDFromProviderID(node.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+
+	matchingSubnet, matchingPort, err := o.findAssignSubnetAndPort(ip, node)
+	if err != nil {
+		return err
+	}
+
+	// 2) Reserve the IP address on the subnet by creating a new unattached neutron port.
```

Comment:
Do the numbers in comments still make sense in `AssignPrivateIP` I don't think we need them with the introduction of `findAssignSubnetAndPort`